File: src/leadforge/scrapers/thumbtack.py
import structlog
from leadforge.scrapers.base import BaseAPIClient

logger = structlog.get_logger()

class ThumbtackClient(BaseAPIClient):
    """Thumbtack business scraper."""

    def __init__(self):
        super().__init__(base_url="https://www.thumbtack.com", timeout=30.0)

    async def search_business(self, name: str, zip_code: str) -> dict | None:
        """Search for a business on Thumbtack. Returns profile data or None."""
        try:
            client = await self._get_client()
            response = await client.get(
                "/s/",
                params={"search_term": name, "zip_code": zip_code},
                headers={"Accept": "text/html"},
            )
            response.raise_for_status()
            text = response.text

            import re
            import json

            # Try to extract structured data from page
            script_match = re.search(r'<script[^>]*type="application/ld\+json"[^>]*>(.*?)</script>', text, re.DOTALL)
            if script_match:
                try:
                    ld_json = json.loads(script_match.group(1))
                    return {
                        "thumbtack_name": ld_json.get("name"),
                        "thumbtack_rating": ld_json.get("aggregateRating", {}).get("ratingValue"),
                        "thumbtack_review_count": ld_json.get("aggregateRating", {}).get("reviewCount"),
                        "thumbtack_hires": None,  # Not in LD+JSON, needs deeper parsing
                    }
                except json.JSONDecodeError:
                    pass

            return None
        except Exception as e:
            logger.warning("thumbtack_search_failed", error=str(e), name=name)
            return None

File: api/src/scrapers/thumbtack.ts
// =py scrapers/thumbtack
import { aggregateRating, encodeParams, extractLdJson, fetchText } from './base';

export interface ThumbtackProfile {
  thumbtack_name: string | null;
  thumbtack_rating: number | null;
  thumbtack_review_count: number | null;
  thumbtack_hires: number | null;
}

// =py search_business — LD+JSON carries no hire count, so thumbtack_hires is always null and the
// enrichment caller's `if tt_data.get("thumbtack_hires")` guard never fires.
export async function searchBusiness(name: string, zipCode: string): Promise<ThumbtackProfile | null> {
  try {
    const params = encodeParams({ search_term: name, zip_code: zipCode });
    const html = await fetchText(`https://www.thumbtack.com/s/?${params}`, { headers: { Accept: 'text/html' } });

    const ldJson = extractLdJson(html);
    if (!ldJson) return null;

    const rating = aggregateRating(ldJson);
    return {
      thumbtack_name: (ldJson.name as string) ?? null,
      thumbtack_rating: (rating.ratingValue as number) ?? null,
      thumbtack_review_count: (rating.reviewCount as number) ?? null,
      thumbtack_hires: null, // Not in LD+JSON, needs deeper parsing
    };
  } catch (error) {
    console.warn('thumbtack_search_failed', { name, error: error instanceof Error ? error.message : String(error) });
    return null;
  }
}
